### ISSUE TITLE
init-pipeline fails with "unexpected error"

### ISSUE THREAD

**Issue by petecliff** [org-formation/org-formation-cli#181]
Hello,

First let me say this project is fantastic and fills a gap in AWS org setup perfectly! Thank you!

However, I've run into an issue (which I imagine is something I'm doing wrong!) and wondered if you knew a fix.

I'm working through the AWS blogs and at part 2, running:

`org-formation init-pipeline organization.yaml --profile PROFILE --region eu-west-1`

results in the following error:

```INFO: uploading initial commit to S3 organization-formation-ACC_NUM/initial-commit.zip...
ERROR: unexpected error occurred...
Missing credentials in config, if using AWS_CONFIG_FILE, set AWS_SDK_LOAD_CONFIG=1
Error: connect EHOSTDOWN 169.254.169.254:80 - Local (x.x.x.x:x)
    at internalConnect (net.js:907:16)
    at defaultTriggerAsyncIdScope (internal/async_hooks.js:301:12)
    at net.js:998:9
    at processTicksAndRejections (internal/process/task_queues.js:75:11)
```
    
This is on a new AWS account, running an IAM user with full admin and MFA enabled. I can upload to that bucket with the same profile using aws cli and have also tried setting AWS_SDK_LOAD_CONFIG etc.

Thanks again!

**Comment by zaro0508** [comment]
what version of node and org-formation are you using?  have you tried just the `init` command (org-formation init --region eu-west-1)?

**Comment by OlafConijn** [comment]
and... what type of mechanism do you use to load your credentials for this profile? i believe some of the newer once are not fully supported by the sdk i use. looking forward to learn more about this.

a somewhat ugly workaround could be to load the access_key and secret over the command.

thanks

**Comment by eduardomourar** [comment]
Possibly related to #123

**Comment by petecliff** [comment]
Hello - thanks for getting back.

Node v12.13.0 and org-formation v0.9.14

Credentials are from .aws/credentials file using --profile - I have tried both directly with the user and also assuming a role as per the note on MFA in the README.

`org-formation init` works - no error and the org file is created.

I'll try to dig into this some more today to see where it might be falling over. Thanks!

**Comment by petecliff** [comment]
I did a bit of digging and it looks like this is failing as `this.s3credentials` is set to undefined in `init-organization-pipeline.ts`.

I got the command working in two ways:

First it works if I create a role in the org master account called `OrganizationAccountAccessRole` and specify the `--build-account-id` parameter with the master account id set. I cannot use any other role name here and it looks to be hardcoded on line 19 of `aws-util.ts`.

Second, I got it working by changing `init-organization-pipeline.ts` to:
```
        command.delegateToBuildAccount = command.buildAccountId !== undefined;
        if (command.delegateToBuildAccount) {
            command.buildProcessRoleName = 'OrganizationFormationBuildAccessRole';
            this.buildAccountId = command.buildAccountId;
            this.s3credentials = await AwsUtil.GetCredentials(command.buildAccountId, DEFAULT_ROLE_FOR_CROSS_ACCOUNT_ACCESS.RoleName);
        } else {
            this.s3credentials = await AwsUtil.GetCredentials(this.buildAccountId, DEFAULT_ROLE_FOR_CROSS_ACCOUNT_ACCESS.RoleName);
        }
```
and then ignore the `useCurrentPrincipal` boolean in `aws-util.ts` which means `s3credentials` are set and the upload to S3 can proceed.

I'm telling you that in the hope it helps - not because I think it is a viable code change! :-)

**Comment by OlafConijn** [comment]
madness ~ i think we got thrown on the wrong track here a bit.

The `--build-account-id` was intended to be used only if the build-account-id would be different from the master account id.
I could imagine that causing all sorts of weird errors

if the build does not run in the master account another set of roles needs to be provisioned for everything to work well
- BuildAccount -> MasterAccount access will be done using the OrganizationFormationBuildRole.
- BuildAccount -> Account Creation will be done using OrganizationFormationBuildRole and then uses OrganizationAccountAccessRole to set up some defaults.
- BuildAccount -> AnyOtherAccount  access will be done using the OrganizationFormationBuildRole.

all these default role names can be changed.

if the current user runs within the master account it is ok for the s3credentials to be undefined, assuming the user has enough permissions to interact with the s3 bucket. if this leads to a permission issue, the user will just need more perms.

i believe the following would fix your issue: 
1) not specifying the --build-account-id flag
2) ignoring it if this is the same as the master account id  (like here: https://github.com/org-formation/org-formation-cli/commit/247fec4ec6cd4f9925528862b52f21db344c44c7)

**Comment by petecliff** [comment]
Yep - I was using this as an excuse to look at/play with the code and didn't think that my hacks would be a good option.

If I do not specify the --build-account-id flag, I get the access denied error as above even with latest. The bucket *is* created but it is empty. I turned off MFA just to reduce variables.

**Comment by petecliff** [comment]
I finally got to testing the workaround and that works without code changes... Had to assume the role via CLI and then set environment variables for session creds.

**Comment by OlafConijn** [comment]
we have made some improvements to how credentials get resolved.
solving this issue for now, looking forward to issues people encounter on 0.9.15-beta11 (and eventually 0.9.15 and up).

thanks

**Comment by josel1204** [comment]
> I finally got to testing the workaround and that works without code changes... Had to assume the role via CLI and then set environment variables for session creds.

How did you do that, could you please provide the steps that you followed. 
Thanks!!

**Comment by OlafConijn** [comment]
@josel1204, do you have issues when running init-pipeline?

**Comment by josel1204** [comment]
Hello @OlafConijn, I was running into some issues and I found the root cause. May be useful for anyone else (maybe not), I set up my profile within ~.aws/credentials but I left a blank space at the beginning (copy/paste issue) but was difficult to figure it out because it was creating the bucket without issues and it wasn't able to upload the zip file it was giving Access Denied.

Thanks for replying to my comment.
